lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/ClosingBinder.java

```diff
@@ -0,0 +1,110 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.base;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.io.Closer;
+import com.google.inject.Binder;
+import com.google.inject.BindingAnnotation;
+import com.google.inject.Inject;
+import com.google.inject.Key;
+import com.google.inject.Scopes;
+import com.google.inject.multibindings.Multibinder;
+import jakarta.annotation.PreDestroy;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.lang.annotation.Annotation;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+
+import static com.google.inject.multibindings.Multibinder.newSetBinder;
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+public class ClosingBinder
+{
+    public static ClosingBinder closingBinder(Binder binder)
+    {
+        return new ClosingBinder(binder);
+    }
+
+    private final Multibinder<ExecutorService> executors;
+    private final Multibinder<Closeable> closeables;
+
+    public ClosingBinder(Binder binder)
+    {
+        executors = newSetBinder(binder, ExecutorService.class, ForCleanup.class);
+        closeables = newSetBinder(binder, Closeable.class, ForCleanup.class);
+        binder.bind(Cleanup.class).in(Scopes.SINGLETON);
+    }
+
+    public ClosingBinder registerExecutor(Class<? extends ExecutorService> type)
+    {
+        executors.addBinding().to(type);
+        return this;
+    }
+
+    public ClosingBinder registerExecutor(Class<? extends ExecutorService> type, Class<? extends Annotation> annotation)
+    {
+        executors.addBinding().to(Key.get(type, annotation));
+        return this;
+    }
+
+    public ClosingBinder registerCloseable(Class<? extends Closeable> type)
+    {
+        closeables.addBinding().to(type);
+        return this;
+    }
+
+    public ClosingBinder registerCloseable(Class<? extends Closeable> type, Class<? extends Annotation> annotation)
+    {
+        closeables.addBinding().to(Key.get(type, annotation));
+        return this;
+    }
+
+    private static class Cleanup
+    {
+        private final Set<ExecutorService> executors;
+        private final Set<Closeable> closeables;
+
+        @Inject
+        public Cleanup(
+                @ForCleanup Set<ExecutorService> executors,
+                @ForCleanup Set<Closeable> closeables)
+        {
+            this.executors = ImmutableSet.copyOf(executors);
+            this.closeables = ImmutableSet.copyOf(closeables);
+        }
+
+        @PreDestroy
+        public void shutdown()
+                throws IOException
+        {
+            executors.forEach(ExecutorService::shutdownNow);
```

Comment:
I'm not sure if the `shutdownNow` is the preferred behaviour. I can imagine that some folks might want to implement graceful shutdown with a timeout?

Do we try to call the 'shutdown' or 'close' methods on all the objects even if the first one in the loop throws an exception?

Reply:
I think it is desired behavior. At this point when we stop the lifecycle all the processing should be done or abandoned. Also this code fallows how it was already implemented in places where there were no resource leaks. I prefer to change one thing at a time.

Reply:
We could make this more generic and explicit by having a single registration type
```java
 public ClosingBinder register(Class<?> type, Callable callable) ...

 public ClosingBinder register(Class<?> type, Class<? extends Annotation> annotation, Callable callable) ...
```
Then the callers would provide the action
```java
.register(ExecutorService.class, ForAsyncHttp.class, ExecutorService::shutdownNow)
```

Reply:
I have tried to make it generic and I am unable to implement in a simple way. This generic methods requires a possibility to store a different closure method for a resource instance. To implement this I need to store some kind of mapping of a resource instance to its closure method. 

The current implementation is simple as executor service will be closed only in a single way and closeable is also closed in single way.